ui/src/funding/pool.ts

```diff
@@ -0,0 +1,335 @@
+export type Weight = BigNumberish;
+
+export enum ReceiverStatus {
+  // The receiver is being added
+  Added = "Added",
+  // The receiver is being removed
+  Removed = "Removed",
+  // The receiver is present already
+  Present = "Present",
+}
+
+export function make(wallet: Wallet): Pool {
+  const data = remote.createStore<PoolData>();
+  const poolContract = contract.pool(wallet.signer);
+  const erc20TokenContract = contract.erc20Token(wallet.signer);
+  loadPoolData();
+
+  // Periodically refresh the pool data. Particularly useful to
+  // reactively display incoming support made available to the user,
+  // update the displayed pool remaining balance, etc.
+  const POLL_INTERVAL_MILLIS = 10000;
+  setInterval(() => {
+    loadPoolData();
+  }, POLL_INTERVAL_MILLIS);
+
+  async function loadPoolData() {
+    if (svelteStore.get(wallet).status !== WalletStatus.Connected) return;
+
+    try {
+      const balance = await poolContract.withdrawable();
+      const collectableFunds = await poolContract.collectable();
+      const amountPerBlock = await poolContract.amountPerBlock();
+      const contractReceivers = await poolContract.receivers();
+      const receivers = new Map<Address, ReceiverStatus>(
+        contractReceivers.map((e: contract.PoolReceiver) => [
+          e.receiver,
+          ReceiverStatus.Present,
+        ])
+      );
+      const erc20Allowance = await getErc20Allowance();
+
+      data.success({
+        // Handle potential overflow using BN.js
+        balance: Number(balance),
+        amountPerBlock: amountPerBlock.toString(),
+        receivers,
+        // Handle potential overflow using BN.js
+        collectableFunds: Number(collectableFunds),
+        erc20Allowance,
+      });
+    } catch (error) {
+      data.error(error);
+    }
+  }
+
+  function getAccount(): Account | undefined {
+    const w = svelteStore.get(wallet);
+    return w.status === WalletStatus.Connected
+      ? w.connected.account
+      : undefined;
+  }
+
+  async function onboard(
+    topUp: BigNumberish,
+    amountPerBlock: BigNumberish,
+    receivers: Receivers
+  ): Promise<void> {
+    return poolContract
+      .onboard(topUp, amountPerBlock, toReceiverWeights(receivers))
+      .then((tx: ContractTransaction) => {
+        transaction.add(
+          transaction.supportOnboarding(tx, topUp, amountPerBlock, receivers)
+        );
+        tx.wait();
+      })
+      .finally(loadPoolData);
+  }
+
+  async function updateSettings(
+    amountPerBlock: string,
+    receivers: Receivers
+  ): Promise<void> {
+    return poolContract
+      .updatePlan(amountPerBlock, toReceiverWeights(receivers))
+      .then((tx: ContractTransaction) => {
+        const currentReceivers = data.unwrap()?.receivers || new Map();
+        const newReceivers = newSetOfReceivers(currentReceivers, receivers);
+        transaction.add(
+          transaction.updateSupport(tx, amountPerBlock, newReceivers)
+        );
+        tx.wait();
```

Comment:
Aren’t we missing an `await` or `return` here? Otherwise `finally` will be called _before_ the transaction has been submitted.

Reply:
Where exactly? I don't think that's the case that `finally` is called before submission but I will check!

Reply:
Verified that `finally` only runs when the transaction is either approved or rejected in the wallet.

Reply:
> Verified that `finally` only runs when the transaction is either approved or rejected in the wallet.

That’s correct. The promise returned from `.updatePlan()` only resolves when the transaction has been submitted to the chain successfully. But why do we call `tx.wait()` here and don’t wait for it?

Reply:
Ah, I now get your question :) I believe this is out you have set this up initially, so I just followed the lead here. I am not sure if we need to `.wait` at all since we will be looking up the statuses of ongoing transactions ya. What do you recon?

Reply:
> I am not sure if we need to `.wait` at all since we will be looking up the statuses of ongoing transactions ye.

I don’t think we need to `.wait()` either. Let’s just remove this line.

Reply:
Done!